### PR TITLE
Hotp token 2fa

### PR DESCRIPTION
**Pull request by erudenko**

### CODE REVIEW

File: web/api/2fa.go [hunk 1]
```diff
@@ -169,6 +180,21 @@ func (ar *Router) FinalizeTFA() http.HandlerFunc {
...
 
+func (ar *Router) verifyOTPCode(user model.User, otp string) (bool, error) {
+	result := false
+	if ar.tfaType == model.TFATypeApp {
+		totp := gotp.NewDefaultTOTP(user.TFAInfo().Secret)
+		result = totp.Verify(otp, int(time.Now().Unix()))
+	} else {
+		hotp := gotp.NewDefaultHOTP(user.TFAInfo().Secret)
+		result = hotp.Verify(otp, user.TFAInfo().HOTPCounter)
+		if user.TFAInfo().HOTPExpiredAt.Before(time.Now()) {
```
Comment (Compaurum) [inline web/api/2fa.go:191]: I don't understand how `user.TFAInfo` works but I think it's better to check Expire time firstly, and then do other checks

File: web/api/token_middleware.go [hunk 2]
```diff
@@ -19,6 +18,8 @@ const (
 	TokenTypeAccess = "access"
 	// TokenTypeRefresh is a refresh token type.
 	TokenTypeRefresh = "refresh"
+	// TokenTypeTFAPreauth is an 2fa preauth token type.
+	TokenTypeTFAPreauth = "2fa-preauth"
```
Comment (Compaurum) [inline web/api/token_middleware.go:22]: why do we need these duplications of tokens? The same here `jwt/service/token_service.go`

File: web/api/2fa.go (hunk 1 above)
Comment (Compaurum) [inline web/api/2fa.go:191]: I don't understand how user.TFAInfo works but I think it's better to check Expire time firstly, and then do other checks

File: cmd/demo/server-config.yaml [hunk 3]
```diff
@@ -33,7 +33,7 @@ configurationStorage:
   settingsKey: server-config.yaml
   keyStorage: # Storage for keys used for signing and verifying JWTs.
     type: local # Key storage type. Supported values are "local" and "s3".
-    folder: ../../jwt # Folder for public.pem and private.pem keys. Assumed to be root if ommitted.
+    folder: ../../test # Folder for public.pem and private.pem keys. Assumed to be root if ommitted.
```
Comment (Compaurum) [inline cmd/demo/server-config.yaml:36]: it's not the best name for keys storage

**Review by Compaurum — CHANGES_REQUESTED** [review]

File: .vscode/launch.json [hunk 4]
```diff
@@ -43,6 +43,25 @@
...
+            "request": "launch",
+            "mode": "debug",
+            "remotePath": "",
+            "port": 2345,
+            "host": "127.0.0.1",
+            "program": "${workspaceFolder}/cmd/demo/main.go",
+            "cwd": "${workspaceFolder}/cmd/demo/",
+            "env": {
+                "SES_REGION": "us-east-1",
+                "SES_SENDER": "noreply@cco.link",
```
Comment (esimonov) [inline .vscode/launch.json:58]: cco.link?

**Review by esimonov — CHANGES_REQUESTED** [review]

File: web/api/token_middleware.go (hunk 2 above)
Comment (SuddenGunter) [reply web/api/token_middleware.go:22]: agree with @Compaurum 
I would suggest not to introduce new token type, but grant limited scope for this token. Like limit role or, well `scope: [tfa]` array in token payload. So the user with this token could only go to one endpoint

Comment (erudenko) [reply web/api/token_middleware.go:22]: @Compaurum this is a historical issue, I will fix it.

Comment (erudenko) [reply web/api/token_middleware.go:22]: @SuddenGunter amazing proposal, I will do the scope instead.

File: .vscode/launch.json (hunk 4 above)
Comment (erudenko) [reply .vscode/launch.json:58]: creator connect leftovers :-)

**Review by esimonov — APPROVED** [review]

**Review by SuddenGunter — APPROVED** [review]

**Pull request merged**